### PR TITLE
🔥 Do not detaches sockets / terminates client stream

### LINKED ISSUE #2170: Connection closed before response was received
### Package

dio

### Version

5.4.2+1

### Operating-System

Android, iOS

### Adapter

Default Dio

### Output of `flutter doctor -v`

```shell
[✓] Flutter (Channel stable, 3.19.4, on macOS 14.4.1 23E224 darwin-arm64, locale en-EG)
[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2023.2)
[✓] VS Code (version 1.87.2)
[✓] Connected device (3 available)            
    ! Error: Browsing on the local area network for Me. Ensure the device is unlocked and attached with a cable or associated with the same
      local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)
    ! Error: Browsing on the local area network for Mohamed’s Iphone (2). Ensure the device is unlocked and attached with a cable or
      associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)
    ! Error: Browsing on the local area network for ‏iPhone ‏Fadwa. Ensure the device is unlocked and attached with a cable or associated
      with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)
    ! Error: Browsing on the local area network for H. Sayed. Ensure the device is unlocked and attached with a cable or associated with the
      same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)
[✓] Network resources

• No issues found!
```

### Dart Version

3.3.2 

### Steps to Reproduce

1. Run with any version of Dio above 5.4.0.
2. try to send more requests in a short period of time. 

### Expected Result

When using Dio version 5.4.0, the connection is not closed and the response succeeds.

untitled2

### Actual Result

flutter: The following exception was thrown getBOKRechargeRequestStatusProvider:
flutter: AppException.serverException(type: ServerExceptionType.unknown, message: HttpException: Connection closed before response was received, uri = https://api.bravo-sudan.com/bravo/api/v1/bok/getRechargeRequests, code: null)
flutter: 
#0      DioMixin.fetch (package:dio/src/dio_mixin.dart:509:7)
<asynchronous suspension>
#1      MainApiFacade._errorHandler (package:bravo/core/infrastructure/network/main_api/api_callers/main_api_facade.dart:66:14)
<asynchronous suspension>
#2      TransactionRemoteDataSource.getBOKRechargeRequestStatus (package:bravo/features/transaction/infrastructure/data_sourse/transaction_remote_remote_data_source.dart:48:20)
<asynchronous suspension>
#3      TransactionHistoryRepository.getBOKRechargeRequestStatus (package:bravo/features/transaction/infrastructure/repository/transaction_history_repository.dart:43:20)

untitled

### PR DESCRIPTION
**Pull request by AlexV525**
Fixes #2170

### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests without failures
- [x] I have updated the `CHANGELOG.md` in the corresponding package

### CODE REVIEW

**Comment by AlexV525** [comment]
I've failed to write tests for these changes. I've tried using regular implementation and also Mockito, the `detachSocket` or `terminate` happens after the current event loop of the request. Now we stop calling them means no innovations. And the exception requires some duration gap which I cannot find during my tests writing.

**Comment by AlexV525** [comment]
I'm probably running into the state that some contents are using the main ref. The test works well locally.

**Comment by kuhnroyal** [comment]
Looks reasonable, we should find a reproducible test case before we merge this imo.

**Comment by AlexV525** [comment]
> Looks reasonable, we should find a reproducible test case before we merge this imo.

I'd appreciate any ideas. I still cannot find a solid case unless we can make some integration tests.

**Comment by AlexV525** [comment]
Oh babe, coding is such a poetic act that can only be solved if you have feelings.

I've added relevant tests and all passed locally.

| dio | http2 |
|-----|-----|
| image | image |

File: dio_test/lib/src/test/cancellation_tests.dart [hunk 1]
```diff
@@ -113,5 +113,33 @@ void cancellationTests(
...
+        );
+        // Simulate connection established.
+        await Future.delayed(const Duration(seconds: 1));
+        token.cancel();
+        final response = await completer.future;
+        // Response should be obtained without exceptions.
+        expect(response, isNotNull);
+      },
+      testOn: '!browser',
+    );
```
Comment (kuhnroyal) [inline dio_test/lib/src/test/cancellation_tests.dart:143]: Nice!
I think this is easier to read without completer.

```suggestion
    test(
      'not closing sockets with requests that have same hosts #2170',
      () async {
        final token = CancelToken();

        await expectLater(
          dio.get('/get', cancelToken: token),
          completes,
        );

        expectLater(
          dio.get('/drip?duration=3'),
          completes,
        );

        Future.delayed(const Duration(seconds: 1), () {
          token.cancel('cancelled');
        });
      },
      testOn: '!browser',
    );
```

**Review by kuhnroyal — APPROVED** [review]

**Pull request merged**